### ISSUE TITLE
Proposal: Add "functions" to templates

### ISSUE THREAD

**Issue by technosophos** [helm/helm#2084]
I've been trying to come up with a way to call templates like functions (with arguments).

So far, the best I've been able to do is:

```
{{ include "some.template" (dict "NAME" "value") }}
```

Would it be helpful to simulate a Bash-like function call like this?

```
# Execute a template as if it were a function that takes arguments.
{{ func "some.template" $param $param2 }}
```

This could be used as a way to call templates that look like this:

```
{{ define "some.template" }}
  # Arguments are passed in using an Args array.
  {{ $arg1 := index .Args 0 }}
  {{ $arg2 := index .Args 1 }}
  # Do something
{{ end }}
```

**Comment by thenamli** [comment]
This is something I am looking for and I was planning to read gotpl in detail for finding this functionality. 

Current state of templates is getting repetitive and I believe this will solve it

**Comment by alanmeadows** [comment]
We are using something like your example:

```
{{ define "some.template" }}
  # Arguments are passed in using an Args array.
  {{ $arg1 := index .Args 0 }}
  {{ $arg2 := index .Args 1 }}
  # Do something
{{ end }}
```

See https://github.com/att-comdev/openstack-helm/blob/master/helm-toolkit/templates/_endpoints.tpl#L129

My issue with this is it lacks support for optional arguments, which is important if we think of charts as an evolving piece of software.  Everything is positional which makes refactoring hard, and it makes optional arguments impossible because there is no good way to gracefully test for the existence of elements that I am aware of.  This is also applicable to accessing nested elements where the parent doesn't exist implodes just about everything.

In other words

```
{{- define "helm-toolkit.my_function" -}}
{{- $arg1 := index . 0 -}}
{{- $arg2 := index . 1 -}}
{{- $optional_arg3 := index . 3 -}}
```

In the example above, if arg3 is not passed, at least as an empty string (which is not ideal for readability) the index call will implode.

The ability to silently test for the presence of vars, lengths of arrays, and be able to access non-existent nested maps without exploding will really facilitate using what already exists as function building blocks.

**Comment by technosophos** [comment]
@alanmeadows Thanks for the feedback. I'll have to see to what extent I can accommodate optional params without risking the behavior of a failed `index` lookup. I wonder if there is a way to cleverly wrap a `define` block in a way that I could intercept some parameters.

**Comment by databus23** [comment]
How about passing the arguments as a map to the template instead of a array.

e.g. `map[string]string{"1":"value1", "2":"value2"}`

Shouldn't this then work?
```
{{- $arg1 := .1 -}}
{{- $arg2 := .2 -}}
{{- $optional_arg3 := .3 | default "None" -}}
```

**Comment by technosophos** [comment]
@databus23 That could work well. I like that idea.

**Comment by sgnl19** [comment]
We're using a dict to inject parameters for readability reasons:
```
{{- define "some.template" -}}
{{- $arg1 := index . "arg1" -}}
{{- $arg2 := index . "arg2" -}}
...
{{- end -}}

{{ some.template (dict "arg1" "first arg" "arg2" "second arg")
```

**Comment by bacongobbler** [comment]
I believe this was addressed via `tpl`: https://helm.sh/docs/howto/charts_tips_and_tricks/#using-the-tpl-function

Closing, but please let me know if I'm wrong and we can reopen this.
